### ISSUE TITLE
Should the NULL cipher be included in the FIPS provider?

### ISSUE THREAD

**Issue by jwalch** [openssl/openssl#13618]
I noticed that deflt_ciphers[] in the default provider includes ossl_null_functions at the top, but the FIPS provider does not have it in fips_ciphers. It wasn't entirely clear to me that the omission of the NULL cipher was an intentional choice. I looked a bit into the history of the fips_ciphers array, and it seems like it was always more of an "opt in" thing rather than having been an "opt out" exercise relative to the default provider. If I am mistaken and this was previous discussed by the OTC or during some code review and the decision to omit is intentional, a brief comment in fips_ciphers may still be useful.

As to the question itself, I can kind of see arguments for either answer. For my money, the decision to encrypt or not encrypt is a subtly distinct one from the decision of how to encrypt and I think choosing not to encrypt is still fundamentally valid even while in a FIPS-compliant mode of operation. It's possible client code may wish to share common paths for "encrypt" and "don't encrypt" for whatever reason.

As to the specific context that led me down this road, I am working on a project that involves integrating OpenSSH and OpenSSL. I noticed that after doing "default_properties = fips=yes" in the openssl.cfg that I was hitting failures in ssh_packet_set_connection (https://github.com/openssh/openssh-portable/blob/a5ab499bd2644b4026596fc2cb24a744fa310666/packet.c#L293) when trying to SSH in from a client (ie, while the code in question was operating in server mode). As we can see, OpenSSH is using a NULL cipher here. My assumption is that this is some kind of early setup thing that happens prior to the SSH protocol itself forming agreement on things like Kx, cipher, MAC, etc. But I'm not an OpenSSH expert, so maybe I'm wrong about that.

At any rate, if I added the NULL cipher to the beginning of the list is fips_ciphers[], things seemed to start working smoothly. (I would note that our project has made some modifications to stock OpenSSH in the past and so it's possible stock wouldn't work with a "FIPsy" openssl.cnf anyway.)

**Comment by jwalch** [comment]
As my reasoning may seem incomplete, a bit of clarification...

When I say client code may wish to share common paths, I am referring to the usage of the EVP_CIPHER(_CTX) family of APIs (Decrypt, Encrypt, Cipher + Init, Update, Final). It may be possible to special-case the NULL cipher to be handled outside of the provider (or in the base or default provider) in all these places, but my assumption is that as a matter of practical concern it would simply be easier to include the NULL cipher in the FOM.

FWIW, I took a look at the old FOM, and it seems like it did include the NULL cipher: https://github.com/openssl/openssl/commit/9ebc37e667704041f9d31f916f41e7d2e0b3feb0

**Comment by richsalz** [comment]
The FIPS provider should only have FIPS-approved ciphers, and NULL isn't one of them.

**Comment by jwalch** [comment]
**shrug** As obvious as perhaps that airtight syllogism may be to YOU, it evidently does not reflect the design decisions made in prior renditions of FIPS support for the project...

**Comment by richsalz** [comment]
The primary people involved in the 1.0.2-based FOM are not part of the project any more.

**Comment by jwalch** [comment]
Why do you suppose that matters?

Inclusion of the NULL cipher in the 1.0.2-based FOM was rather obviously an intentional decision. Either it was the wrong one all along, or something meaningful has changed between now and then. Obviously it would be convenient to be able to have the individual responsible provide relevant information that didn't make it into the change history or code comments, but it shouldn't be *necessary* in order to be able to sensibly evaluate things.

Furthermore, the fact that it happened in the first place - whether the individuals responsible for it are still associated with the project or not - is a point of evidence that your terse statement is not as obviously true to everyone as you seem to think it is. Someone committed the change, perhaps someone else even reviewed it, and no one went in and explicitly corrected it.

Obviously it is the case that the NULL cipher is explicitly prohibited by the guidance in certain contexts, such as within the broader concept of TLS cipher suites. However, it's not like every bit written to disk, memory, wire, etc. is mandated to be encrypted; heck, even TLS itself as a protocol needs a bit of plaintext at the beginning to be able to work. I have trouble wrapping my head around why conceptually the NULL cipher is, from a FIPS-compliance standpoint, meaningfully distinct from not running the data in question through any cipher at all.

I am trying very hard to remain open to the idea that the one-word answer to the question I pose is "no" for various implementation-specific reasons. Perhaps the FIPS provider can't expose the NULL cipher because then it would start permitting things it shouldn't (eg, TLS_RSA_WITH_NULL_SHA or some other similarly poor suite). However, your initial answer struck me as an egregiously insufficient one. I think the existing data strongly suggests the situation has considerably more nuance to it than you are trying to represent it as having.

**Comment by richsalz** [comment]
I believe it matters because if non of the earlier designers of the 1.0.2 FOM are around, then we don't know why NULL was included.  We do know why NULL isn't included in 3.0: because it's not a FIPS cipher.

**Comment by jwalch** [comment]
> We do know why NULL isn't included in 3.0 [...]

Perhaps you might, but "we" don't. I was just trying to sort it out from the commit history and code comments and I certainly didn't see that reason given. I appreciate that you are now providing that reason, but it has no basis in the "public record" that I have been able to discern.

As to why I find the reason itself ("it's not a FIPS cipher") insufficient, it's that the statement is such a patently obvious one and it would have been in the past too. While we can say that we don't know why it was there in the past, I think we have to assume that the past committers knew full well that "NULL isn't a FIPS cipher" was true. So if they knew that and came to a different conclusion ANYWAY, they must have had some other information available to them. Maybe we don't know what that particular information was, but its mere existence would suggest that maybe, just MAYBE, even today we should try to consider the issue more deeply than just leaving it at "it's not a FIPS cipher, done."

I think a big source of confusion here is that employing the "Null Object" design pattern has created a situation in which the semantic meaning of the object is subtly overloaded. Imagining an OpenSSL world in which EVP_enc_null() didn't exist at all (regardless of provider), the the EVP_CIPHER(_CTX) objects and their family of APIs would basically be answering the question of how the data is encrypted; their mere existence would already imply that it's encrypted somehow. However, because EVP_enc_null() is a thing, they also speak to the question of whether the data is encrypted at all.

This is where to my mind there is a gap between how the FIPS guidance defines things and how OpenSSL's APIs realize concepts. The FIPS guidance presupposes the answer to the initial question of whether the data even needs to be encrypted at all is "yes" and then speaks to the natural follow-up question of "how" from the POV of cryptographic robustness.

**Comment by richsalz** [comment]
Did you read https://www.openssl.org/docs/OpenSSL300Design.html ?  In particular this part:
```
FIPS Module

This is a FIPS 140-2 validated cryptographic module. It is a provider that contains FIPS validated/approved cryptographic algorithms only. Non FIPS algorithms will be supplied by the default provider (not the FIPS module).
```

Maybe also read the FIPS 140 standards?  (That is not easy going, especially when you add in the Implementation Guidance.)

**Comment by jwalch** [comment]
Again, I think the ambiguity here is in whether for these purposes the OpenSSL concept of the NULL cipher is even really a "cryptographic algorithm" or in the NIST parlance a "security function."

It seems like your interpretation of the NULL cipher in OpenSSL is that it's essentially the weakest possible cryptographic function. I certainly wouldn't argue that interpretation is unambiguously wrong, and there are definitely situations in which it's definitely the only sensible one. The argument I am trying to make is that I believe there also exists a reasonable alternative interpretation, which is that the NULL cipher isn't even an "algorithm" or "function" at all, but rather something OpenSSL exposes as a convenience mechanism.

Let's move away from the specific OpenSSH problem that prompted me to ask the question, and instead consider an application that provides a user some mechanism that generates data and writes it out to a file. Suppose that mechanism provides an *optional* parameter for the user to specify an encryption algorithm. If the parameter isn't specified, the data is written out as plaintext.

Now, the application COULD only pipe the data through the EVP_CIPHER family of APIs if the optional parameter is specified. However, from the client's code hygiene perspective, it keeps the control flow considerably simpler if you just always go through them; you have some local EVP_CIPHER* that you initialize with EVP_enc_null() and then re-populate it using one of the lookup functions if the user specified an algorithm. Suppose the application code is written the second way, then. Without any provider queries, it works just fine.

Where we run into trouble, of course, is if we're running against a "FIPsy" openssl.cnf; the optional parameter now becomes effectively required, which is probably not desirable. I think there is a legitimate use case for applications wanting to only use FIPS-validated cryptographic implementations if cryptography is to be used at all, but not to mandate cryptography be used at all.

So now to make this work the client has to go through some extra gymnastics (admittedly not THAT bad from a code standpoint, since I believe a simple EVP_CIPHER_fetch() should suffice) relative to what worked perfectly fine in OpenSSL 1.0.2. I get that the FIPS module was re-designed, but it's frustrating because now in this particular regard the application actually has to be *more* aware of FIPS-mode specific semantics than they were before. That doesn't feel like an outcome compatible with any kind of design goals around improved usability.

**Comment by richsalz** [comment]
Note that i was with the project not during the 1.0.2 FIPS days, and I left awhile ago. I represent my employer as a sponsor of the FIPS 3.0 work.  Perhaps others from the OpenSSL project want to comment.  Speaking for myself, I can't consider the "NULL cipher" a security function.

**Comment by paulidale** [comment]
I wouldn't consider the NULL cipher a security function either.  I'm not entirely comfortable with having it in the FIPS provider but if it were, I'd have expect it to have the "fips=no" property present.

I think this is one for the OTC to discuss.

I'm also not convinced that this counts as a regression: FIPS is all new in 3.0, there is no backwards to be compatible with.

**Comment by levitte** [comment]
We have allowed unapproved things in the FIPS provider module already, with the property "fips=no".  For the moment, I only see one, the test RNG.  That has been judged acceptable.
(We should probably review the stuff currently available in the FIPS provider, to see that we got all the "fips=yes" and "fips=no" right, and that we have a common understanding of what that property actually means...  opinions have fluctuated over time)
